Select OEM/ANSI code page according to system locale setting

unxed opened the pull request:
Fixes

https://sourceforge.net/p/sevenzip/bugs/2473/
https://sourceforge.net/p/sevenzip/bugs/1060/

unxed committed:
Select OEM/ANSI code page according to system locale setting

unxed committed:
Apply fix from https://salsa.debian.org/debian/7zip/-/merge_requests/14

LupinThidr commented:
Hello, thank you for this. I've seen your progress across multiple issue trackers regarding Linux and 7zip path encoding.
I ran across your comments while trying to diagnose 7zip unable to decode Shift-JIS (CP932) encoded paths in LZH files
I had thought that this patch would apply to those, but it seems ZipItem is specific to the .zip handler. 
Other software such as unar don't fully support the .LZH spec as 7zip does.
I think your iconv conversion should be implemented as a separate function in StringConvert, so it can be easily used with other classes.
I was able to make LzhHandler use it rather the standard MultiByteToUnicodeString, but I don't completely understand the 7zip codebase, so I've hardcoded it to CP932 rather than the a mcp argument.

```cpp
void MultiByteToUnicodeString3_iconv(UString &res, const AString &s)
{
  res.Empty();
  if (s.IsEmpty())
    return;

  iconv_t cd;
  if ((cd = iconv_open("UTF-8", "CP932")) != (iconv_t)-1) {

    AString sUtf8;

    unsigned slen = s.Len();
    char* src = s.Ptr_non_const();

    unsigned dlen = slen * 4 + 1; // (source length * 4) + null termination
    char* dst = sUtf8.GetBuf_SetEnd(dlen);
    const char* dstStart = dst;

    memset(dst, 0, dlen);

    size_t slen_size_t = static_cast<size_t>(slen);
    size_t dlen_size_t = static_cast<size_t>(dlen);
    size_t done = iconv(cd, &src, &slen_size_t, &dst, &dlen_size_t);

    if (done == (size_t)-1) {
      iconv_close(cd);

      // iconv failed. Falling back to default behavior
      MultiByteToUnicodeString2(res, s, 932);
      return;
    }

    // Null-terminate the result
    *dst = '\0';

    iconv_close(cd);

    AString sUtf8CorrectLength;
    size_t dstCorrectLength = dst - dstStart;
    sUtf8CorrectLength.SetFrom(sUtf8, static_cast<unsigned>(dstCorrectLength));
    if (ConvertUTF8ToUnicode(sUtf8CorrectLength, res) /*|| ignore_Utf8_Errors*/)
      return;
  }

}
```

and then in LzhHandler's GetProperty:

```cpp
UString dst;
MultiByteToUnicodeString3_iconv(dst, item.GetName());

UString s = NItemName::WinPathToOsPath(dst);
```

If you'd like to try implementing it for LZH, here's a sample file and the current / expected output of ``7zz l``
https://archive.org/download/narcissu/na_sabun.lzh
```
2007-06-03 22:25:18 .....          537          392  na_sabun/▒C▒▒▒▒▒▒.txt
```

```
2007-06-03 22:25:18 .....          537          392  na_sabun/修正差分.txt
```

unxed commented:
@LupinThidr can you please provide sample LZH file with file names encoded in CP932, and also list of the same file names in UTF-8 format for reference?

unxed commented:
> I don't completely understand the 7zip codebase

Unfortunately, I also don't understand the 7-zip codebase well enough. I'm afraid we need Igor Pavlov to tell us how to implement this.

Neustradamus commented:
@ip7z: What do you think about this PR?

kattjevfel commented:
This and the patch applied to the debian package completely wrecks shift_jis encoded filenames.
normal 7zip: `�k�����s[mp4]`
with patch: `ûkÅ­Ä×ìs[mp4]`

Difference is that the first one can be fixed with `convmv -f shift_jis -t utf8 -r <path>`, but the second cannot, as has already improperly been converted to UTF-8. FWIW this always happened with p7zip, but was fixed with 7zip (and now re-broken, hooray)

unxed commented:
Will look at this, thanks!

kattjevfel commented:
Absolutely, here you go.
[test.zip](https://github.com/user-attachments/files/18412723/test.zip)

without patch: `20240323�����C�C�`���C�`�����������C�h/���p�K���i�ŏ��ɂ��ǂ݉�����.]Read this first�j.TXT`
with patch: `20240323é¿òùÿCâCâ`âââCâ`ââÅ¼ê½ûéâüâCâh/ùÿùpïKû±üiì┼Åëé╔é¿ôÃé¦ë║é│éó.]Read this firstüj.TXT`
convmv'd: `20240323お風呂イチャイチャ小悪魔メイド/利用規約（最初にお読み下さい.]Read this first）.TXT`

Also fwiw the current output with this patch is the same I get in Ark (KDE's archive GUI), no matter if I have an unpatched 7zip. But that's probably them just relying on how p7zip worked.

Unzipping with pure 7z or specifying encoding with unzip is currently the only way to get these filenames without corrupting them.

unxed commented:
What is your system locale setting? Please provide `locale` command output if possible.

unxed commented:
@kattjevfel please try to replace
```
    // Detect required code page name from current locale 
    char *lc = setlocale(LC_CTYPE, "");

    if (lc && lc[0]) {
```
to
```
    // Detect required code page name from current locale 
    char *lc = setlocale(LC_CTYPE, "");
    if (!lc || !lc[0]) {
      lc = getenv("LC_CTYPE");
    }

    if (lc && lc[0]) {
```

after applying this patch, then check you archive once again. Make sure your system locale is set to ja_JP as 7zip should somehow "guess" what legacy code page to use, so its using system locale for that.

kattjevfel commented:
Your patch does work after adding `#include <cstdlib>`, though having to run it with `LANG=ja_JP.UTF-8` every time I encounter a zip packaged on a Japanese system is kind of annoying.

At that point I still prefer having them not be converted to UTF-8 by 7-zip at all and instead have an external tool identify and convert them, as that doesn't even require the locale to be installed.

unxed commented:
After change above it should also work without locale installed.

kattjevfel commented:
You are right, when using `LANG=ja_JP.UTF-8` without the locale installed, it behaves like without the patch entirely, which is at least better than the current state.

unxed commented:
Use `LC_CTYPE=ja_JP.UTF-8` to enable patch

unxed committed:
Merge branch 'ip7z:main' into main

unxed committed:
do not break things if locale is not installed

unxed committed:
added missing include

unxed committed:
fix length calculation

unxed commented:
Fix applied to Debian version also:
https://salsa.debian.org/debian/7zip/-/merge_requests/15

defrag257 commented:
this code is wrong which can cause some bugs:
```
char *lc = setlocale(LC_CTYPE, "");
```
in fact `setlocale(LC_CTYPE, "something or empty string")` returns the previous LC_CTYPE in `lc` as the man page for setlocale(3). the correct way is using `setlocale(LC_CTYPE, NULL)` as following:
```
setlocale(LC_CTYPE, "");
char *lc = setlocale(LC_CTYPE, NULL);
```
however, using setlocale may interfere with some other codes as it is a global status for c runtime library (msvcrt or glibc etc.), and you may mimic the behavior of setlocale for the priority: `LC_ALL` > `LC_CTYPE` > `LANG`.

defrag257 commented:
Sorry for the previously sent wrong information about setlocale(3) (I have deleted it).

defrag257 commented:
```cpp
// Detect required code page name from current locale
char *lc = setlocale(LC_CTYPE, "");
if (!lc || !lc[0]) {
  lc = getenv("LC_CTYPE");
}
```
should be:
```cpp
// Detect required code page name from current locale
char *lc = setlocale(LC_CTYPE, "");
if (!lc || !lc[0]) {
  lc = getenv("LC_ALL");
}
if (!lc || !lc[0]) {
  lc = getenv("LC_CTYPE");
}
if (!lc || !lc[0]) {
  lc = getenv("LANG");
}
```
or if you do not want to change the global selocale status:
```cpp
// Detect required code page name from current locale
char *lc = getenv("LC_ALL");
if (!lc || !lc[0]) {
  lc = getenv("LC_CTYPE");
}
if (!lc || !lc[0]) {
  lc = getenv("LANG");
}
```

unxed commented:
Thanks! Will apply this change today later

unxed committed:
fix by @defrag257

unxed commented:
@defrag257 pls check now

nhz2 commented:
I checked in <https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT>, and the behavior implemented in this PR is not correct according to the .ZIP specification.

Please follow the specification by default.

Also, this behavior doesn't make sense. The computer creating a .ZIP file may not be the same computer used to extract the file. The interpretation of the zip entry names is just not related to the system locale settings on the computer extracting the file.

Maybe this behavior can be optionally enabled by a command-line switch? But IMO it is not worth the added complexity and added `iconv` dependency.

unxed commented:
But it's just a way Windows XP and some others do it

unxed commented:
The most popular zip archives creator in the world

nhz2 commented:
Maybe you can create a tool called something like "utf8myzip" that will fix up the encoding of a .ZIP file. As input it takes the nonportable .ZIP file and any information on the system and program that created that file. Then as output it creates a new .ZIP file with the UTF-8 flag set, and UTF-8 encoded entry names. People who want to deal with nonportable .ZIP files could first use that tool, and people who want to follow the spec can just use 7zip directly.

unxed commented:
The thing is, a huge number of archivers on Windows use the Windows approach for compatibility. We just have to follow their example. Yes, it’s not a perfect solution, but the sender and receiver are speaking the same language with a probability very close to 1. There’s also unar, which tries to detect the encoding based on the frequency of certain characters or something like that. The risk of incorrect detection there is much higher than the chance that the sender and receiver don’t use the same language.

Also: libarchive outputs raw data as is, and without additional processing this results in invalid UTF-8 codes in the file system. Doesn’t look like an elegant solution.

nhz2 commented:
Windows seems to have fixed this bug. When I create a zip archive using the file explorer now, I get this new ui

<img width="906" height="257" alt="image" src="https://github.com/user-attachments/assets/d3c56992-fcdd-473e-a07d-a3bd0a77dc7f" />

And the created archive is marked as from Unix, and UTF-8 encoded:

```
Name: Å.txt
Folder: -
Size: 2
Packed Size: 4
Modified: 2025-11-06 21:08:40
Attributes:  -rw-rw-rw-
Encrypted: -
CRC: D8932AAC
Method: Deflate
Characteristics: ux UT:M:1 : Descriptor UTF8
Host OS: Unix
Version: 20
Volume Index: 0
Offset: 70
------------------------: 
Name: enctest\
Size: 2
Packed Size: 4
Folders: 0
Files: 1
CRC: D8932AAC
------------------------: 
Path: C:\Users\nzimm\testfiles\enctest.zip
Type: zip
Physical Size: 352
------------------------: 
------------------------: 
```


Also in terms of invalid UTF-8 codes in the file system, the way the most linux filesystems work, this is not actually that bad, and since no data is lost, the filenames can be fix up after extracting, if needed.

gyurix reviewed (commented):
I would not merge this as-is.

This copies huge locale->codepage tables into `ZipItem.cpp`, duplicates external mapping data, and depends on `iconv` code-page names like `CP850` / `CP1252` that are not portable across iconv implementations. That is a lot of policy/data pushed into a hot archive-decoding path with no tests.

Code quality concern: env-var parsing + charset mapping + conversion logic all land in one low-level function.

Merge readiness: 3/10. Before merge: move mapping to a dedicated helper/data source, define fallback behavior precisely, and add cross-locale test cases before changing ZIP filename decoding.

Neustradamus commented:
@unxed: Have you progressed on your PR?

unxed commented:
Thanks for the review! I'll make all the necessary edits as soon as I have some time. I'll try to do it quickly.

unxed committed:
Merge branch 'ip7z:main' into main

unxed committed:
Reimplement legacy OEM/ANSI codepage detection on UNIX platforms for ZIP and LZH archives.

This implementation replaces previous version by centralizing the locale-to-codepage mapping logic within the core StringConvert module. This approach improves portability by using standard iconv encoding names, and provides a fallback mechanism to the default 7-Zip conversion behavior if the requested charset is unavailable. Additionally, this commit introduces a standalone unit test harness, TEST_LegacyMapping.cpp, to verify the cross-locale codepage mapping logic independently of the full archive processing pipeline.

unxed commented:
Check now, please!

unxed commented:
> copies huge locale->codepage tables

I'm copying these tables not because it's architecturally beautiful, but because Wine (the original source of this data) didn't separate them into a standalone library that could be used outside Wine. [libnatspec](https://github.com/Etersoft/libnatspec) did this independently, but it is almost absent from distributions, either because of bias towards Russian developers (although this is strange, I'm Russian and my patches were accepted by many projects), or God knows why else.

PS: Also, at the moment, the Wine and libnatspec tables differ. I assumed Wine is better source of truth, as it's being developed more actively, but that's not necessarily correct.

CC @Etersoft
